src/database-controller/src/write-merger/handler.js

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+const createError = require('http-errors');
+const logger = require('@dbc/core/logger');
+const AsyncLock = require('async-lock');
+const DatabaseModel = require('openpaidbsdk');
+const config = require('@dbc/write-merger/config');
+const {
+  Snapshot,
+  AddOns,
+  silentSynchronizeRequest,
+} = require('@dbc/core/framework');
+const _ = require('lodash');
+const lock = new AsyncLock({ maxPending: Number.MAX_SAFE_INTEGER });
+const databaseModel = new DatabaseModel(
+  config.dbConnectionStr,
+  config.maxDatabaseConnection,
+);
+
+/* For error handling, all handlers follow the same structure:
+
+  try{
+    ....
+  } catch (err) {
+    return next(err)
+  }
+
+  If a normal HTTP error is wanted, use `createError`.
+  e.g. return next(createError(401, 'Please login to view this page.'))
+
+  If a 500 error is wanted, throw it in-place: throw new Error('fatal error')
+
+*/
+
+async function ping(req, res, next) {
+  try {
+    res.status(200).json({ message: 'ok' });
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function receiveWatchEvents(req, res, next) {
+  try {
+    const snapshot = new Snapshot(req.body);
+    const frameworkName = snapshot.getName();
+    if (!frameworkName) {
+      return next(createError(400, 'Cannot find framework name.'));
+    }
+    await lock.acquire(frameworkName, async () => {
+      const oldFramework = await databaseModel.Framework.findOne({
+        attributes: ['snapshot'],
+        where: { name: frameworkName },
+      });
+      // database doesn't have the corresponding framework.
+      if (!oldFramework) {
+        if (config.recoveryModeEnabled) {
+          // If database doesn't have the corresponding framework,
+          // and recovery mode is enabled
+          // tolerate the error and create framework in database.
+          const record = snapshot.getRecordForLegacyTransfer();
+          record.requestSynced = true;
+          await databaseModel.Framework.create(record);
+        } else {
+          throw createError(404, `Cannot find framework ${frameworkName}.`);
+        }
+      } else {
+        // Database has the corresponding framework.
+        const oldSnapshot = new Snapshot(oldFramework.snapshot);
+        const internalUpdate = {};
+        if (oldSnapshot.getGeneration() === snapshot.getGeneration()) {
+          // if framework request is equal, mark requestSynced = true
+          logger.info(`The request of framework ${frameworkName} is synced.`);
+          internalUpdate.requestSynced = true;
+        } else {
+          // if framework request is not equal,
+          // should use framework request in db as ground truth
+          internalUpdate.requestSynced = false;
+        }
+        // use request in database
+        snapshot.overrideRequest(oldSnapshot);
+        if (req.params.eventType === 'DELETED') {
+          // if event is DELETED, mark apiServerDeleted = true
```

Comment:
req.params.eventType === 'DELETED' && framework is completed

Reply:
So if running state framework is unexpected deleted directly in ApiServer by some DRIs, DB will still create it in ApiServer again, instead of let the framework stuck in running state.

---
In reply to: [466222182](https://github.com/microsoft/pai/pull/4761#discussion_r466222182) [](ancestors = 466222182)

Reply:
Below is better:
```
       if (req.params.eventType === 'DELETED') {
           if framework is completed
               internalUpdate.apiServerDeleted = true;
           else
               update framework status to be the mocked one
        }
```

---
In reply to: [466223160](https://github.com/microsoft/pai/pull/4761#discussion_r466223160) [](ancestors = 466223160,466222182)

Reply:
if (req.params.eventType === 'DELETED') {
          if (snapshot.getState() === 'Completed') {
            // if event is DELETED and the state is Completed, mark apiServerDeleted = true
            internalUpdate.apiServerDeleted = true;
          } else {
            // Event is DELETED and the state is not Completed.
            // This case could occur when someone deletes the framework in API server directly.
            // In such case, we mark requestSynced=false, and reset the snapshot using the framework request.
            snapshot = new Snapshot(snapshot.getRequest(false))
            internalUpdate.requestSynced = false;
          }
        }

Reply:
ok